crates/revm/src/optimism.rs

```diff
@@ -0,0 +1,113 @@
+            })
+            .and_then(|tx| {
+                tx.input().get(4..).ok_or(executor::BlockExecutionError::L1BlockInfoError {
+                    message: "could not get l1 block info tx calldata bytes".to_string(),
+                })
+            })?;
+
+        // The setL1BlockValues tx calldata must be exactly 184 bytes long, considering that
+        // we already removed the first 4 bytes (the function selector). Detailed breakdown:
+        //   8  bytes for the block number
+        // + 8  bytes for the block timestamp
+        // + 32 bytes for the base fee
+        // + 32 bytes for the block hash
+        // + 8  bytes for the block sequence number
+        // + 32 bytes for the batcher hash
+        // + 32 bytes for the fee overhead
+        // + 32 bytes for the fee scalar
+        if l1_info_tx_data.len() != 184 {
+            return Err(executor::BlockExecutionError::L1BlockInfoError {
+                message: "unexpected l1 block info tx calldata length found".to_string(),
+            })
+        }
+
+        let l1_base_fee = U256::try_from_le_slice(&l1_info_tx_data[16..48]).ok_or(
+            executor::BlockExecutionError::L1BlockInfoError {
+                message: "could not convert l1 base fee".to_string(),
+            },
+        )?;
+        let l1_fee_overhead = U256::try_from_le_slice(&l1_info_tx_data[120..152]).ok_or(
+            executor::BlockExecutionError::L1BlockInfoError {
+                message: "could not convert l1 fee overhead".to_string(),
+            },
+        )?;
+        let l1_fee_scalar = U256::try_from_le_slice(&l1_info_tx_data[152..184]).ok_or(
+            executor::BlockExecutionError::L1BlockInfoError {
+                message: "could not convert l1 fee scalar".to_string(),
+            },
+        )?;
+
+        Ok(Self { l1_base_fee, l1_fee_overhead, l1_fee_scalar })
+    }
+}
+
+impl L1BlockInfo {
+    /// Calculate the gas cost of a transaction based on L1 block data posted on L2
+    pub fn calculate_tx_l1_cost(&mut self, tx: &TransactionSigned) -> U256 {
+        let rollup_data_gas_cost = U256::from(tx.input().iter().fold(0, |acc, byte| {
+            acc + if *byte == 0x00 { ZERO_BYTE_COST } else { NON_ZERO_BYTE_COST }
+        }));
+
+        if tx.is_deposit() || rollup_data_gas_cost == U256::ZERO {
+            return U256::ZERO
+        }
+
+        rollup_data_gas_cost
+            .saturating_add(self.l1_fee_overhead)
+            .saturating_mul(self.l1_base_fee)
+            .saturating_mul(self.l1_fee_scalar)
+            .checked_div(U256::from(1_000_000))
+            .unwrap_or_default()
+    }
+}
+
+/// Get the base fee recipient address
+pub fn base_fee_recipient() -> Address {
+    Address::from_str(BASE_FEE_RECIPIENT).unwrap()
+}
+
+/// Get the L1 cost recipient address
+pub fn l1_cost_recipient() -> Address {
+    Address::from_str(L1_FEE_RECIPIENT).unwrap()
+}
```

Comment:
better to change constants to statics wrapped in `once_cell::sync::Lazy` rather than parsing each time this is called. in that case no need for these functions either